broker/lib/bosh/BoshOperationCache.js

```diff
@@ -0,0 +1,261 @@
+    logger.info('Removing task from cache for service instance', serviceInstanceId);
+    const wrapper = new Promise((resolve, reject) => {
+      const key = getTaskKey(serviceInstanceId);
+      etcdConnector().delete().key(key).then(obj => {
+        resolve(obj);
+      }).catch(err => reject(err));
+    });
+    return wrapper;
+  }
+
+  /**
+   * Deletes a single BOSH deployment from the Etcd Store
+   * 
+   * @param {string} deploymentName - Deployment Name
+   */
+  deleteDeploymentFromCache(deploymentName) {
+    logger.info('Removing deployment from cache', deploymentName);
+    const wrapper = new Promise((resolve, reject) => {
+      etcdConnector().delete().key(getKey(deploymentName)).then(out => resolve(out)).catch(err => reject(err));
+    });
+    return wrapper;
+  }
+
+  /**
+   * Deletes multiple BOSH deployment names from the Etcd store
+   * 
+   * @param {(string|string[])} deployments - List of Deployment Names
+   */
+  deleteDeploymentsFromCache(...deployments) {
+    const deploymentIdentifiers = _.flattenDeep(deployments);
+    logger.info('Removing multiple deployments from cache', deployments);
+    const deletePromises = _.map(deploymentIdentifiers, this.deleteDeploymentFromCache);
+    return Promise.all(deletePromises);
+  }
+
+  /**
+   * Stores the BOSH task ID corresponding to the service instance ID when the task is submitted to the BOSH Director
+   * 
+   * @param {string} serviceInstanceId  - Service Instance ID
+   * @param {string} taskId - BOSH Task ID corresponding to the operation
+   */
+  storeBoshTask(serviceInstanceId, taskId) {
+    logger.info('Putting the current deployment task into the cache', serviceInstanceId, taskId);
+    const wrapper = new Promise((resolve, reject) => {
+      const key = getTaskKey(serviceInstanceId);
+      etcdConnector().put(key).value(taskId)
+        .then(() => resolve(true))
+        .catch(err => {
+          logger.error('Error in storing BOSH task id for instance', err);
+          reject(err);
+        });
+    });
+    return wrapper;
+  }
+
+  /**
+   * Stores the deployment operation (user-triggered) into the Etcd store as a JSON document
+   * Key is prefixed with a specific format and the deployment name is used as the identifier
+   * 
+   * @param {string} planId - Service plan ID for the service instance
+   * @param {string} deploymentName - Calculated bosh deployment name
+   * @param {Object} params - Parameters used for the operation
+   * @param {Object} args - Arguments used for the operation
+   */
+  store(planId, deploymentName, params, args) {
+    logger.info('Putting the current deployment operation into the cache', planId, deploymentName, params, args);
+    const wrapper = new Promise((resolve, reject) => {
+      const key = getKey(deploymentName);
+      const operation = new DeploymentOperation(planId, deploymentName, params, args);
+      this.containsDeployment(deploymentName).then(exists => {
+        if (exists) {
```

Comment:
The operations that go through these checks are only for user-based ops. These ops which always create or update a BOSH deployment are treated asynchronously from the PoV of CC. Yes, they always return `202-ACCEPTED`

Reply:
For the point on decoupling from CC, I would assume the proxy would still remain the same way since the concept of asynchronous provisioning/updates should not change. Of course, I would treat this for future refactoring if required